typed-racket-lib/typed-racket/typecheck/tc-structs.rkt

```diff
@@ -348,44 +277,100 @@
       (let ([variance (hash-ref (free-vars-hash (free-vars* t)) var variance:const)])
         (or (variance:const? variance)
             (and (not mutable) (variance:co? variance))))))
+
   (define covariant?
     (and (covariant-for? self-fields mutable)
          (covariant-for? parent-fields parent-mutable)))
 
   (define (poly-wrapper t) (make-Poly tvars t))
+  (mk-bindings! sty names desc si
+                (lambda (st-type-alias)
+                  (if (not covariant?)
+                      (make-StructTop sty)
+                      (subst-all (make-simple-substitution tvars (map (const Univ) tvars)) st-type-alias)))
+                (lambda (name self-fld idx-parent-cnt st-type-alias)
+                  (let* ([path (make-StructPE st-type-alias idx-parent-cnt)]
+                         [func (poly-wrapper
+                                (if mutable
+                                    (->* (list st-type-alias) self-fld)
+                                    (->acc (list st-type-alias) self-fld (list path))))])
+                    (values (list st-type-alias idx-parent-cnt mutable #f) func)))
+                (lambda (_ self-fld idx-parent-cnt st-type-alias)
+                  (values (list st-type-alias idx-parent-cnt #f) (poly-wrapper (->* (list st-type-alias self-fld) -Void))))))
+
+;; if val-th is not false, evaluate it and cons the result onto rst
+;; (-> (Option (-> A) (Listof A) (Listof A))
```

Comment:
I don't understand why the thunking is useful here.

Reply:
https://github.com/racket/typed-racket/pull/1091/files#diff-516cab1093154106286b0cff961a6e54c2468f64a452f70f4bb8a47b73e82df3R369

In this case, semantically, if mutators don't exist at all, we don't need to compute an empty list?

Reply:
You are right. Thunking is unnecessary